spv-in parse more atomic ops

schell opened the pull request:
**Connections**
Part of
* https://github.com/gfx-rs/wgpu/issues/4489

Depends on
* https://github.com/gfx-rs/wgpu/pull/5775

**Description**

This adds parsing for the following atomic operations to `naga::spv::Frontend::next_block`: 

* [x] OpAtomicLoad
* [x] OpAtomicStore
* [x] OpAtomicExchange
* [ ] OpAtomicCompareExchange 
* [ ] ~OpAtomicCompareExchangeWeak~ (deprecated) 
* [ ] ~OpAtomicIIncrement~ was included in #5702
* [x] OpAtomicIDecrement
* [x] OpAtomicIAdd 
* [x] OpAtomicISub 
* [x] OpAtomicSMin 
* [x] OpAtomicUMin 
* [x] OpAtomicSMax 
* [x] OpAtomicUMax 
* [x] OpAtomicAnd 
* [x] OpAtomicOr 
* [x] OpAtomicXor 
* [x] OpAtomicFlagTestAndSet 
* [x] OpAtomicFlagClear


**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

schell committed:
add parsing for spirv::Op::AtomicLoad and spirv::Op::AtomicStore

schell committed:
spv-in parse AtomicExchange and AtomicCompareExchange

schell committed:
add atomic i decrement

schell committed:
bookend atomic store statement with emmitter.finish/emitter.start to suppress a double load expression

bookend atomic result expressions with emitter.finish/start to prevent double defs

schell committed:
add atomic iadd, isub, smin, umin, smax, umax, and, or, xor

schell committed:
parse atomic flag test and set, parse atomic flag clear

schell committed:
remove atomic compare exchange work

schell committed:
changelog

jimblandy reviewed (commented):
The tests in `src/front/atomic_upgrade.rs` really belong in `src/front/spv`. We can't have SPIR-V specific code outside `src/front/spv` and `src/back/spv`.

jimblandy commented on naga/src/front/atomic_upgrade.rs:
```diff
@@ -212,3 +212,82 @@ impl Module {
+#[cfg(test)]
+mod test {
+    fn atomic_test(bytes: &[u8]) {
+        let _ = env_logger::builder().is_test(true).try_init();
```
Wouldn't it make more sense to `.expect` this `Result` than ignore it?

schell replied:
I would say yes if this were `main`, where one would expect that logging would only be set up once. But in a test it may be that the logger has already been initialized by some other test, in which case I don't want to panic. 

Usually when using logging in a test I'll use `ctor` to ensure that we only init once, but that would turn on logging for all tests and introduce a new dependency.

jimblandy replied:
Oh, right. I forgot about all that mess.

schell committed:
moved spirv tests into front/spv/mod.rs

jimblandy commented on naga/src/front/spv/mod.rs:
```diff
@@ -5710,32 +5964,44 @@ mod test {
+                    should_panic = true;
+                }
+                Ok(i) => {
+                    wgsl = crate::back::wgsl::write_string(
```
This won't compile if you only build with the `spv-in` feature, and not the `wgsl-out` feature, right? I think this function needs to be conditional on more cfg values. Note that this stuff changed recently; see the `cfg_aliases!` macro invocation in `naga/build.rs` for what you should use now.

jimblandy commented on naga/src/front/spv/mod.rs:
```diff
@@ -5710,32 +5964,44 @@ mod test {
-    #[test]
-    fn atomic_i_inc() {
+    fn atomic_test(bytes: &[u8]) {
         let _ = env_logger::builder().is_test(true).try_init();
```
Okay, this `Result` needs to be ignored because `env_logger` treats multiple initialization attempts as an error, right? I hope that won't stifle other errors. But I don't see anything better that can be done here.

schell replied:
Yes, exactly.

schell commented:
@teoxoy do you think you would feel comfortable giving this a look over?

schell committed:
feature gate atomic spv tests because they require wgsl-[in,out]

schell commented:
@jimblandy do you have any more comments? I think we're getting pretty close with this 🤞 .

jimblandy commented on naga/src/front/spv/mod.rs:
```diff
@@ -1323,6 +1323,94 @@ impl<I: Iterator<Item = u32>> Frontend<I> {
+        block.push(stmt, span);
+
+        // Store any associated global variables so we can upgrade their types later
+        self.upgrade_atomics
+            .extend(ctx.get_contained_global_variable(p_lexp_handle));
```
Each time the SPV front end calls `get_contained_global_variable`, we use `extend` to add the global to `Frontend::upgrade_atomics`, This means that if `get_contained_global_variable` can't identify the global to which this atomic is referring, we don't upgrade anything. Is this the correct behavior? Why aren't these calls returning an error?

schell replied:
Previously this function recursed and returned `IndexSet<Handle<crate::GlobalVariable>>`, but then I changed that to `Option<Handle<crate::GlobalVariable>>` as per your suggestion. But I think it makes sense to error, as we'll always be dealing with a global variable.

jimblandy replied:
Without pretending to understand the consequences, just in case, here's a commit that makes it an error if we can't identify a global to be upgraded: https://github.com/gfx-rs/wgpu/commit/52d43fd5c245001585d8a0253c8d845e76e6ba7b

jimblandy committed:
Generate spans covering the entire instruction.

Granted, there is pre-existing code in the SPIR-V front end that gets
this wrong, but:

It doesn't make sense to read `self.data_offset`, and then immediately
pass that to `self.span_from_with_op`. The point of that function is
to make the span cover the entire instruction, operands included.

jimblandy commented:
@schell Does this commit look okay? https://github.com/gfx-rs/wgpu/commit/87b01090af5363d77fb54292d8a94dc60f8a1e3b

schell commented:
@jimblandy that commit seems just fine 👍

schell committed:
BlockContext::get_contained_global_variable returns Result

jimblandy commented:
@schell So, right now this branch has a bunch of merges from trunk. We are generally trying to merge PRs by rebasing, because we don't want the work-in-progress commits in our history forever. Could you change this PR's branch to be merge-free?

Edit: I believe the standard `git rebase` behavior will just drop merges. Then you can fixup/squash as appropriate. cc our git authority: @ErichDonGubler

jimblandy committed:
Move `From` implementation into spv front end

jimblandy committed:
doc comments, minor cleanups

schell commented:
@jimblandy I rebased earlier today, so I think we're good to go - though now it's out of date again, so I'll do that again.

schell commented:
I think I also need to cherry-pick your spans commit?

schell commented:
Ok - my rebase earlier clobbered your "span" commit, so I cherry-picked it back in. The change of `get_contained_global_variable` to return a `Result` I made at the same time you did, and your change got clobbered as well, but my change covers that issue. So now I think we're g2g 👍

schell commented:
Tests still pass and everything :)

jimblandy commented on naga/src/front/spv/mod.rs:
```diff
@@ -4027,31 +4192,141 @@ impl<I: Iterator<Item = u32>> Frontend<I> {
+                        _ => crate::AtomicFunction::ExclusiveOr,
+                    },
+                )?,
+                Op::AtomicFlagTestAndSet => {
```
There's no test coverage for this, or for `Op::AtomicFlagClear`.

jimblandy reviewed (changes requested):
I think the flag ops are not going to work, because WGSL doesn't support `atomic<bool>`, so those `Literal::Bool` values are going to make something unhappy.

schell commented:
Ok, I'll add tests for those and debug it. As an alternative do you think we can make it use `u32`?

jimblandy commented:
Yes, treating the argument as `atomic<u32>` or `atomic<i32>` seems consistent with the SPIR-V ops, too, which say that the pointer should be a pointer "to a 32-bit integer type".

schell commented:
Yeah, you're right. I just missed that. Ok, I'll just stick to that then.

jimblandy commented:
If the tests pass with the code as it is, then if I'm reading correctly (and I'm often not), that would mean that the validator is not properly type-checking `Statement::Atomic`, because the pointer would be to an `atomic<i32 | u32>`, but the value would be a `bool`.

schell commented:
You know what - I don't even have the ability to create code that uses the flag ops (not from rust-gpu, at least). I say we skip these until a later date.

schell committed:
remove parsing of OpAtomicFlagClear and OpAtomicFlagTestAndSet

schell committed:
Merge remote-tracking branch 'schell/feature/spirv-front-atomics-3-more-ops' into feature/spirv-front-atomics-3-more-ops

jimblandy reviewed (approved):
Okay, with the untested ops removed, this is good to go.

jimblandy reviewed (changes requested):
Sorry - I just noticed that some of the `.spvasm` files in `naga/tests/in/spv` do not match their `.spv` files. That needs to be squared away before we can merge.

jimblandy commented:
I've filed #6292 as a follow-up for converting the tests to snapshot tests.

schell committed:
sync atomic spvasm files
